<a href="https://colab.research.google.com/github/ncuello95/Algoritmos_de_optimizacion2023/blob/main/Seminario_Algoritmos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Algoritmos de optimización - Seminario<br>
Nombre y Apellidos: Nicolás Emiliano Cuello Oliver  <br>
Url: https://github.com/ncuello95/Algoritmos_de_optimizacion2023<br>
Problema:

>2. Organizar los horarios de partidos de La Liga<br>


Descripción del problema:
Desde la La Liga de fútbol profesional se pretende organizar los horarios de los partidos de
liga de cada jornada. Se conocen algunos datos que nos deben llevar a diseñar un
algoritmo que realice la asignación de los partidos a los horarios de forma que maximice
la audiencia.


(*) La respuesta es obligatoria





                                        

======================================================================================================================================================================================================================================

**(*)¿Cuantas posibilidades hay sin tener en cuenta las restricciones?<br>**

Respuesta

Inicialmente lo que se necesita saber es de cuantas posibles maneras pueden agruparse los 20 equipos de a 2, sin repeticiones.

In [ ]:
import itertools
import math

equipos = ['A1','A2','A3','B1','B2','B3','B4','B5','B6','B7','B8','B9','B10','B11', 'C1','C2','C3','C4','C5','C6']

partidos = list(itertools.combinations(equipos, 2))
contador = 0
lista_partidos = []
for partido in partidos:
    lista_partidos.append(partido)
    contador = contador + 1

print('Se pueden generar:',contador,'posibles combinaciones de partidos')

Se pueden generar: 190 posibles combinaciones de partidos


El primer partido puede elegirse de entre las 190 posibilidades existentes:
A partir del segundo, los posibles partidos ya se encuentran reducidos en 1 (dado el partido elegido anteriormente) y en n*2 siendo 'n' el numero de veces que aparecen los equipos elegidos
 en las posibles combinaciones restantes, y 2 los equipos que se enfrentan.

 Por ej: Si se elige como primer partido (A1,A2)
 Para el segundo partido ya tengo 1 elegido, es decir, se reduce el numero total de 190 a 189
 y luego tengo que observar que los equipos A1 y A2 ya no se encuentran disponibles para el segundo partido, y dado que cada uno de ellos aparece 18 veces mas, es preciso restar esta cantidad.

 Para el resto de partidos se sigue un razonamiento análogo, solo que en lugar de 18 se va reduciendo de a 2.

 El resutado del siguiente cálculo también responde a la siguiente expresión:
`20!/2^10`

In [ ]:
partido1 = 190
print('La primer elección de partidos puede ser de: ',partido1)
partido2 = 190-1-18*2
print('La segunda elección de partidos puede ser de: ',partido2)
partido3 = partido2-1-16*2
print('La tercera elección de partidos puede ser de: ',partido3)
partido4 = partido3-1-14*2
print('La cuarta elección de partidos puede ser de: ',partido4)
partido5 = partido4-1-12*2
print('La quinta elección de partidos puede ser de: ',partido5)
partido6 = partido5-1-10*2
print('La sexta elección de partidos puede ser de: ',partido6)
partido7 = partido6-1-8*2
print('La septima elección de partidos puede ser de: ',partido7)
partido8 = partido7-1-6*2
print('La octava elección de partidos puede ser de: ',partido8)
partido9 = partido8-1-4*2
print('La novena elección de partidos puede ser de: ',partido9)
partido10 = partido9-1-2*2
print('La decima elección de partidos puede ser de: ',partido10)

n = partido1*partido2*partido3*partido4*partido5*partido6*partido7*partido8*partido9*partido10

print('El número de posibles agrupaciones de partidos de a 10 es de:',n)

La primer elección de partidos puede ser de:  190
La segunda elección de partidos puede ser de:  153
La tercera elección de partidos puede ser de:  120
La cuarta elección de partidos puede ser de:  91
La quinta elección de partidos puede ser de:  66
La sexta elección de partidos puede ser de:  45
La septima elección de partidos puede ser de:  28
La octava elección de partidos puede ser de:  15
La novena elección de partidos puede ser de:  6
La decima elección de partidos puede ser de:  1
El número de posibles agrupaciones de partidos de a 10 es de: 2375880867360000


Una vez obtenido este valor, debe tenerse en cuenta como los partidos pueden estar distribuidos. Una posibilidad, sin tener en cuenta ninguna restricción, podría ser elegir el viernes a las 20 para que disputen todos los partidos simultaneamente.

Para conocer este nuevo valor de posibles combinaciones, puede pensarse que al elegir uno de los 10 partidos, cada uno podría ubicarse en cualquiera de los 10 horarios disponibles del fin de semana. Con posibilidades de coincidencia entre fechas (lo que no reduciría el espacio de 10 horarios disponibles).

Esto hace que cada elección deba multiplicarse por 10. Lo que arroja un nuevo número:

In [ ]:
combinaciones_totales_sin_restricciones = n*10*10*10*10*10*10*10*10*10*10
print('Sin tener en cuenta las restricciones existen un total de :',combinaciones_totales_sin_restricciones, 'combinaciones totales')

Sin tener en cuenta las restricciones existen un total de : 23758808673600000000000000 combinaciones totales


Se puede observar que este número responde también al siguiente cálculo:
`20!*10^10 / 2^10`

======================================================================================================================================================================================================================================

**¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones.**

Al tener en cuenta que los viernes y lunes debe jugarse un partido:

In [ ]:
combinaciones_totales_con_restricciones = n*1*8*8*8*8*8*8*8*8*1
print('Sin tener en cuenta las restricciones existen un total de :',combinaciones_totales_con_restricciones, 'combinaciones totales')

Sin tener en cuenta las restricciones existen un total de : 39860666501966069760000 combinaciones totales


Se puede observar que este número responde también al siguiente cálculo:
`20!*8^8 / 2^10`

======================================================================================================================================================================================================================================

 **Modelo para el espacio de soluciones<br>
(*) ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)**


Respuesta

En este caso, la estructura que encuentro mas apta para trabajar con el problema es mediante una estructura de arbol para los partidos sumado a una lista para los horarios. Es decir, en cada nivel del arbol se debe elegir entre uno de los posibles partidos + un horario de la lista: [V20, S12, S16, S18, S20, D12, D16, D18, D20, L20].

La estructura del arbol sería de la siguiente manera:
El primer partido puede elegirse libremente. En cuanto al horario, al ser el primero, ya se encuentra definido: V20.

Para los 8 partidos subsiguientes ya existen limitaciones al reducirse el numero de equipos disponibles.


*   Partido 1: 190 posibles combinaciones + [V20]
*   Partido 2: 153 ramificaciones de posibles combinaciones + un horario de la lista: [S12, S16, S18, S20, D12, D16, D18, D20]
*   Partido 3: 120 ramificaciones de posibles combinaciones + un horario de la lista: [S12, S16, S18, S20, D12, D16, D18, D20]
*   Partido 4: 91 ramificaciones de posibles combinaciones + un horario de la lista: [S12, S16, S18, S20, D12, D16, D18, D20]
*   Partido 5: 66 ramificaciones de posibles combinaciones + un horario de la lista: [S12, S16, S18, S20, D12, D16, D18, D20]
*   Partido 6: 45 ramificaciones de posibles combinaciones + un horario de la lista: [S12, S16, S18, S20, D12, D16, D18, D20]
*   Partido 7: 28 ramificaciones de posibles combinaciones + un horario de la lista: [S12, S16, S18, S20, D12, D16, D18, D20]
*   Partido 8: 15 ramificaciones de posibles combinaciones + un horario de la lista: [S12, S16, S18, S20, D12, D16, D18, D20]
*   Partido 9: 6 ramificaciones de posibles combinaciones + un horario de la lista: [S12, S16, S18, S20, D12, D16, D18, D20]
*   Partido 10: 1 partido restante + [L20]

Habiendo elegido los partidos y horarios para todo el fin de semana. La información se puede visualizar correctamente en un dataframes en pandas, en donde en las columnas se muestra la información de la audiencia estimada junto con los factores de correción.

======================================================================================================================================================================================================================================

**Según el modelo para el espacio de soluciones<br>**
**(*)¿Cual es la función objetivo?**

**(*)¿Es un problema de maximización o minimización?**

Respuesta

La función objetivo se obtiene de la suma de las audiencias estimadas en cada uno de los partidos. Matematicamente puede representarse mediante la siguiente función:

**F_de_costo** = f( [equipo1,equipo2,horario1] , [equipo3,equipo4,horario2] , [equipo5,equipo6,horario3] , [equipo7,equipo8,horario4] , [equipo9,equipo10,horario5] , [equipo11,equipo12,horario6] , [equipo13,equipo14,horario7] , [equipo15,equipo16,horario8] , [equipo17,equipo18,horario9] , [equipo19,equipo20,horario10] )

\
Es un problema de maximización, ya que se trata de maximizar la cantidad de audiencia en los partidos.

======================================================================================================================================================================================================================================

**Diseña un algoritmo para resolver el problema por fuerza bruta**

Respuesta

======================================================================================================================================================================================================================================

**Calcula la complejidad del algoritmo por fuerza bruta**

Respuesta

======================================================================================================================================================================================================================================

**(*)Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta**

Respuesta

La mejora de un algoritmo por fuerza bruta, puede darse por el hecho de aplicar criterios simplificativos con el fin de reducir el espacio de busqueda de las soluciones.

Para este caso los criterios adoptados surgen de la propia observación de las estimaciones de audiencia totales, es decir, de cuando las mismas son máximas y bajo que situaciones. Por ejemplo, se sabe que el partido de equipos de división A: (A1, A2) es el partido que mayor audiencia base genera (2 millones) y que si se programa en el horario de sabado a las 20 o domingo a las 20, este valor resulta multiplicado por un factor de 1. Por lo tanto, resulta sensato, que los partidos que generen mayor audiencia sean colocados en los horarios del fin de semana que también tengan mejor previsión de audiencia.

Sumado a esto, inicialmente se considerará el caso de que no puedan darse coincidencias entre los partidos.

El algoritmo que se desarrolla a continuación se basa en una metodología voraz
en donde para cada día, se elegirá el partido que maximice la audiencia.
El orden de los días ya se encuentra preestablecido en función de los coeficientes de reducción de audiencia:

```
['S20','D20','D18','S18','D16','S16','S12','D12','V20','L20']
```

Posteriormente, se aplicará el algoritmo de recocido simulado para explorar otras zonas de búsqueda para tratar de encontrar otras potenciales zonas con mejores soluciones.

In [2]:
# IMPORT DE LIBRERIAS
import itertools
import math
import pandas as pd
import random
import copy

In [1]:
# DEFINICION DE LAS FUNCIONES A UTILIZAR:
#---------------------------------------------------------
def remover_equipos(lista,eq1,eq2):

    #Se eliminan los equipos ya asignados.
    i = 0
    while i < len(lista):
        if (eq1 in lista[i]) or (eq2 in lista[i]):
            lista.remove(lista[i])
        else:
            i += 1
    return lista
#---------------------------------------------------------
def calcular_audiencia(equipo_1,equipo_2,coeficientes):

    #Se calcula la audiencia.
    if ('A' in equipo_1 and 'A' in equipo_2): audiencia_estimada = 2*coeficientes
    if ('B' in equipo_1 and 'B' in equipo_2): audiencia_estimada = 0.9*coeficientes
    if ('C' in equipo_1 and 'C' in equipo_2): audiencia_estimada = 0.47*coeficientes
    if ('A' in equipo_1 and 'B' in equipo_2) or ('B' in equipo_1 and 'A' in equipo_2): audiencia_estimada = 1.3*coeficientes
    if ('A' in equipo_1 and 'C' in equipo_2) or ('C' in equipo_1 and 'A' in equipo_2): audiencia_estimada = 1*coeficientes
    if ('B' in equipo_1 and 'C' in equipo_2) or ('C' in equipo_1 and 'B' in equipo_2): audiencia_estimada = 0.75*coeficientes

    return audiencia_estimada
#---------------------------------------------------------
def coincidencias(df_voraz):

    #Se genera lista de coeficientes acorde a las coincidencias entre partidos.
    coef_corr = []

    df_coincidencias = pd.DataFrame(data=df_voraz['Horario'].value_counts())
    for i in range(df_coincidencias.shape[0]):
      if df_coincidencias['Horario'][i] == 1: coef_corr.append(1)
      if df_coincidencias['Horario'][i] == 2: coef_corr.append(0.75)
      if df_coincidencias['Horario'][i] == 3: coef_corr.append(0.55)
      if df_coincidencias['Horario'][i] == 4: coef_corr.append(0.4)
      if df_coincidencias['Horario'][i] == 5: coef_corr.append(0.3)
      if df_coincidencias['Horario'][i] == 6: coef_corr.append(0.25)
      if df_coincidencias['Horario'][i] == 7: coef_corr.append(0.22)
      if df_coincidencias['Horario'][i] == 8: coef_corr.append(0.2)
      if df_coincidencias['Horario'][i] == 9: coef_corr.append(0.2)

    df_coincidencias['Coef_de_corr'] = coef_corr
    df_coincidencias = df_coincidencias.reset_index()
    df_coincidencias['Horario'] = df_coincidencias['index']

    return df_coincidencias
#---------------------------------------------------------
def genera_vecina_aleatorio(horario,coeficientes):

    #Se eligen dos nodos aleatoriamente.
    i,j = sorted(random.sample( range(1,len(horario)) , 2))

    #Devuelve una nueva solución pero intercambiando los dos nodos elegidos al azar.
    return horario[:i] + [horario[j]] + horario[i+1:j] + [horario[i]] + horario[j+1:] , coeficientes[:i] + [coeficientes[j]] + coeficientes[i+1:j] + [coeficientes[i]] + coeficientes[j+1:]
#---------------------------------------------------------
def genera_equipos():

    #Lista con la totalidad de los equipos.
    equipos = ['A1','A2','A3','B1','B2','B3','B4','B5','B6','B7','B8','B9','B10','B11', 'C1','C2','C3','C4','C5','C6']

    #Se generan todas las posibles combinaciones, se cuentan y se agregan a la lista: lista_partidos.
    partidos = list(itertools.combinations(equipos, 2))
    contador = 0
    lista_partidos = []

    for partido in partidos:
        lista_partidos.append(partido)
        contador = contador + 1

    return lista_partidos
#---------------------------------------------------------
def generar_partidos_voraz(horarios,coeficiente,lista_partidos):

    #Se generan listas y parámetros iniciales que se utilizaran en el código.
    audiencia_total = 0
    partido_optimo = []
    partido = []
    audiencia_total = []
    audiencia_estimada = []

    #Se recorre un doble bucle, el primero de ellos siendo el de la lista de horarios: ['S20','D20','D18','S18','D16','S16','S12','D12','V20','L20'].
    for i in range(len(horarios)):

      audiencia_max = -1

      #El segundo bucle corresponde a las posibilidades de distintas combinaciones de los equipos para formar los partidos.
      for j in range(len(lista_partidos)):

        #Se elige el primer elemento de la lista 'lista_partidos' y se calcula la audiencia que generaría.
        equipo_1 = lista_partidos[j][0]
        equipo_2 = lista_partidos[j][1]
        audiencia = calcular_audiencia(equipo_1,equipo_2,coeficiente[i])

        #Si la audiencia generada es mayor que la audiencia inicial, o la máxima encontrada hasta el momento se guardan los equipos del partido junto con la nueva audiencia máxima.
        if audiencia > audiencia_max:
          audiencia_max = audiencia
          partido = [equipo_1,equipo_2]

      #Una vez terminado el segundo bucle (de todos las posibles combinaciones de equipos) se guarda el partido óptimo y su audiencia generada.
      partido_optimo.append(partido)
      audiencia_estimada.append(audiencia_max)

      #Los equipos elegidos ya no estarán disponibles para próximas elecciones, por lo tanto se los debe eliminar de la lista_partidos.
      lista_partidos = remover_equipos(lista_partidos,partido[0],partido[1])

      #Se genera una lista en donde cada valor es el máximo de audiencia del horario del primer bucle.
      audiencia_total.append(audiencia_max)

    #Se genera un dataframe en donde por fila se colocan los partidos que generan las mayores audiencias y se agregan las columnas: horario, coeficientes y audiencia estimada.
    df_voraz = pd.DataFrame(partido_optimo, columns=['Equipo1', 'Equipo2'])
    df_voraz['Horario'] = horarios
    df_voraz['Coeficientes'] = coeficiente
    df_voraz['Audiencia Estimada'] = audiencia_estimada

    #El df_coincidencias se genera para contar las posibles coincidencias de horarios entre los partidos (este punto se utilizará en la aplicación del algoritmo con entradas aleatorias).
    #se realiza un merge entre los 2 dataframes para juntar toda la información en un nuevo dataframe llamado df_combinado.
    df_coincidencias = coincidencias(df_voraz)
    df_combinado = pd.merge(df_voraz,df_coincidencias, on='Horario').drop('index', axis=1)
    df_combinado['Correción coincidencia'] = df_combinado['Audiencia Estimada']*df_combinado['Coef_de_corr']

    #El output generado es la suma de la columna 'Corrección coincidencia' que ya tiene la audiencia estimada por partido habiendo aplicado todos los coeficientes necesarios, y el df_combinado
    #que se imprimirá en pantalla.
    output = df_combinado['Correción coincidencia'].sum()

    return output,df_combinado
#---------------------------------------------------------
#Funcion de probabilidad para aceptar peores soluciones
def probabilidad(T,d):
    if random.random() <  math.exp( -1*d / T)  :
      return True
    else:
      return False
#---------------------------------------------------------
#Funcion de descenso de temperatura
def bajar_temperatura(T):
    return T*0.99
#---------------------------------------------------------
def recocido_simulado(TEMPERATURA,lista_partidos,horarios_referencia,coeficientes_referencia):

  lista_partidos_copia = copy.copy(lista_partidos)

  audiencia_referencia,df = generar_partidos_voraz(horarios_referencia,coeficientes_referencia,lista_partidos_copia)

  print('La solución de referencia es: ',horarios_referencia)
  print('La audiencia de referencia es: ', audiencia_referencia)
  print('------------------------------------------')

  mejor_horario = []
  mejor_coeficiente = []
  mejor_audiencia = 0

  N=0
  while TEMPERATURA > .0001:
    N+=1
    #Genera copia de la lista: lista_partidos
    lista_partidos_copia2 = copy.copy(lista_partidos)

    #Genera una solución vecina
    horarios_vecino,coeficiente_vecino = genera_vecina_aleatorio(horarios_referencia,coeficientes_referencia)

    #Calcula el valor de la audiencia vecina
    audiencia_vecina,df = generar_partidos_voraz(horarios_vecino,coeficiente_vecino,lista_partidos_copia2)

    #Si es la mejor solución de todas se guarda(siempre!!!)
    if audiencia_vecina > mejor_audiencia:
        mejor_horario,mejor_coeficiente = horarios_vecino,coeficiente_vecino
        mejor_audiencia = audiencia_vecina
        print('La audiencia vecina número ',N,'es:',mejor_audiencia)

    #Si es peor se cambia según una probabilidad que depende de T y delta(audiencia_referencia - audiencia_vecina)
    if audiencia_vecina < audiencia_referencia or probabilidad(TEMPERATURA, abs(audiencia_referencia - audiencia_vecina) ) :
        horarios_referencia,coeficientes_referencia = horarios_vecino,coeficiente_vecino
        audiencia_referencia = audiencia_vecina

    #Bajamos la temperatura
    TEMPERATURA = bajar_temperatura(TEMPERATURA)

  print("La mejor solución encontrada es " , end="")
  print(mejor_horario)
  print("con una distancia total de " , end="")
  print(mejor_audiencia)

  return mejor_horario,df

Habiendo definido todas las funciones a utilizar, a continuación se realizará el llamado unicamente a las del algoritmo voraz.
Los horarios de referencia a ingresar se encuentran ordenados en orden descendente en base a los coeficientes de reducción de audiencia.

In [3]:
horarios_referencia = ['S20','D20','D18','S18','D16','S16','S12','D12','V20','L20']
coeficientes_referencia = [1,1,0.85,0.8,0.75,0.7,0.55,0.45,0.4,0.4]
lista_partidos = genera_equipos()

#Se aplica el algoritmo voraz con los datos generados anteriormente.
aud_max,df = generar_partidos_voraz(horarios_referencia,coeficientes_referencia,lista_partidos)
print('La aud max es:', aud_max)
df

La aud max es: 7.172499999999999


Equipo1 Equipo2 Horario  Coeficientes  Audiencia Estimada  Coef_de_corr  \
0      A1      A2     S20          1.00              2.0000             1   
1      A3      B1     D20          1.00              1.3000             1   
2      B2      B3     D18          0.85              0.7650             1   
3      B4      B5     S18          0.80              0.7200             1   
4      B6      B7     D16          0.75              0.6750             1   
5      B8      B9     S16          0.70              0.6300             1   
6     B10     B11     S12          0.55              0.4950             1   
7      C1      C2     D12          0.45              0.2115             1   
8      C3      C4     V20          0.40              0.1880             1   
9      C5      C6     L20          0.40              0.1880             1   

   Correción coincidencia  
0                  2.0000  
1                  1.3000  
2                  0.7650  
3                  0.7200  
4                  0.6750  
5                  0.6300  
6                  0.4950  
7                  0.2115  
8                  0.1880  
9                  0.1880

Posteriormente se realizará el llamado a la función de recocido simulado.

Los horarios de referencia a ingresar son los mismos y se encuentran ordenados en orden descendente en base a los coeficientes de reducción de audiencia. Dentro de la función de recocido simulado, se buscarán alternativas mediante la generación de horarios vecinos.

In [5]:
horarios_referencia = ['S20','D20','D18','S18','D16','S16','S12','D12','V20','L20']
coeficientes_referencia = [1,1,0.85,0.8,0.75,0.7,0.55,0.45,0.4,0.4]
lista_partidos = genera_equipos()
Temperatura_ini = 10000000

#Se aplica la función de recocido_simulado con los datos generados anteriormente.
solucion,df  = recocido_simulado(Temperatura_ini,lista_partidos,horarios_referencia,coeficientes_referencia)
df

La solución de referencia es:  ['S20', 'D20', 'D18', 'S18', 'D16', 'S16', 'S12', 'D12', 'V20', 'L20']
La audiencia de referencia es:  7.172499999999999
------------------------------------------
La audiencia vecina número  1 es: 7.0435
La audiencia vecina número  5 es: 7.0649999999999995
La audiencia vecina número  46 es: 7.092499999999999
La audiencia vecina número  178 es: 7.0925
La audiencia vecina número  179 es: 7.1125
La audiencia vecina número  1407 es: 7.129499999999999
La audiencia vecina número  1411 es: 7.1295
La audiencia vecina número  1489 es: 7.1725
La mejor solución encontrada es ['S20', 'D20', 'D18', 'D16', 'S18', 'S16', 'S12', 'L20', 'V20', 'D12']
con una distancia total de 7.1725


Equipo1 Equipo2 Horario  Coeficientes  Audiencia Estimada  Coef_de_corr  \
0      A1      A2     S20          1.00              2.0000             1   
1      A3      B1     S12          0.55              0.7150             1   
2      B2      B3     S16          0.70              0.6300             1   
3      B4      B5     D16          0.75              0.6750             1   
4      B6      B7     V20          0.40              0.3600             1   
5      B8      B9     D12          0.45              0.4050             1   
6     B10     B11     L20          0.40              0.3600             1   
7      C1      C2     D20          1.00              0.4700             1   
8      C3      C4     D18          0.85              0.3995             1   
9      C5      C6     S18          0.80              0.3760             1   

   Correción coincidencia  
0                  2.0000  
1                  0.7150  
2                  0.6300  
3                  0.6750  
4                  0.3600  
5                  0.4050  
6                  0.3600  
7                  0.4700  
8                  0.3995  
9                  0.3760

Lo que se observa es que la aplicación de la metodología voraz, para el estado simplicado del problema, ya arroja una solución que no es mejorable con la aplicación del algorimo metaheurístico. Las soluciones que devuelve resultan ser situaciones no iguales, pero si equivalentes.

======================================================================================================================================================================================================================================

**(*)Calcula la complejidad del algoritmo**

Respuesta

La complejidad del algoritmo diseñado es cuadrática O(n**2).
La función que genera esta complejidad es: 'generar_partidos_voraz' con el doble bucle que ejecuta.

======================================================================================================================================================================================================================================

**Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios**

Respuesta

Para la aplicación del algoritmo diseñado anteriormente se necesita de 3 entradas:


*   Horario, compuesto de dia y hora.
*   Lista de los coeficientes de reducción de audiencia correspondientes.
*   Lista de equipos y posibles combinaciones entre ellos.

A continuación se presentan las funciones que generan estos datos de entrada.

In [6]:
def generar_dia_aleatorio():

    #Se generan días aleatorios entre sabado y domingo. El lunes y viernes se agregarán después ya que, por una restricción del enunciado, no deben repetirse en la lista aleatoria generada
    dias = ['S', 'D']
    dias_aleatorios = random.choice(dias)
    return str(dias_aleatorios)

#-------------------------------------------------------------------
def hora_aleatoria():

    #Se generan horas aleatorias
    hora = [12,16,18,20]
    hora_aleatoria = random.choice(hora)
    return str(hora_aleatoria)

#-------------------------------------------------------------------

def coeficiente_aleatorio(horario_aleatorio):

  #Se genera una lista de coeficientes acorde a los dias y horarios determinados
  coeficiente_aleatorio = []

  for i in range(len(horario_aleatorio)):
    if horario_aleatorio[i] == 'S20': coeficiente_aleatorio.append(1)
    if horario_aleatorio[i] == 'D20': coeficiente_aleatorio.append(1)
    if horario_aleatorio[i] == 'D18': coeficiente_aleatorio.append(0.85)
    if horario_aleatorio[i] == 'S18': coeficiente_aleatorio.append(0.8)
    if horario_aleatorio[i] == 'D16': coeficiente_aleatorio.append(0.75)
    if horario_aleatorio[i] == 'S16': coeficiente_aleatorio.append(0.7)
    if horario_aleatorio[i] == 'S12': coeficiente_aleatorio.append(0.55)
    if horario_aleatorio[i] == 'D12': coeficiente_aleatorio.append(0.45)
    if horario_aleatorio[i] == 'V20': coeficiente_aleatorio.append(0.4)
    if horario_aleatorio[i] == 'L20': coeficiente_aleatorio.append(0.4)

  return coeficiente_aleatorio

#-------------------------------------------------------------------

horario_aleatorio = []

#Bucle para iterar de 0 a 8, ya que los últimos 2 días (viernes y lunes) se agregan al final para que unicamente se encuentren una unica vez.
for i in range(8):
    dia = generar_dia_aleatorio()
    hora = hora_aleatoria()
    horario_aleatorio.append(dia+hora)

horario_aleatorio.append('V20')
horario_aleatorio.append('L20')

horario_aleatorio = random.sample(horario_aleatorio, len(horario_aleatorio))    #la lista final se mezcla aleatoriamente
coeficiente_aleatorio = coeficiente_aleatorio(horario_aleatorio)
equipos_aletaroio = genera_equipos()
lista_partidos_aleatorio = random.sample(equipos_aletaroio, len(equipos_aletaroio)) #la lista final se mezcla aleatoriamente

#-------------------------------------------------------------------

print('Los datos aleatorios generados son:')
print('Horario: ', horario_aleatorio)
print('Coeficientes: ', coeficiente_aleatorio)
print('Equipos: ', lista_partidos_aleatorio)

Los datos aleatorios generados son:
Horario:  ['S12', 'S12', 'D18', 'D16', 'D12', 'D12', 'D18', 'L20', 'V20', 'S20']
Coeficientes:  [0.55, 0.55, 0.85, 0.75, 0.45, 0.45, 0.85, 0.4, 0.4, 1]
Equipos:  [('A1', 'B9'), ('B8', 'C1'), ('A3', 'C3'), ('C3', 'C4'), ('B1', 'C3'), ('B6', 'C2'), ('B6', 'C4'), ('C2', 'C6'), ('C5', 'C6'), ('B1', 'B5'), ('B9', 'C4'), ('C1', 'C5'), ('B1', 'B7'), ('B3', 'C5'), ('B3', 'C6'), ('B11', 'C1'), ('C3', 'C6'), ('A1', 'C4'), ('B1', 'B10'), ('B9', 'B11'), ('B1', 'C1'), ('B1', 'C2'), ('C2', 'C4'), ('B10', 'B11'), ('B2', 'B9'), ('A2', 'C4'), ('B7', 'C4'), ('A3', 'C1'), ('B7', 'C2'), ('B1', 'B4'), ('A3', 'C4'), ('B10', 'C5'), ('A3', 'C6'), ('B6', 'B7'), ('B1', 'C5'), ('A2', 'B2'), ('A1', 'B2'), ('A1', 'C2'), ('B1', 'B2'), ('A3', 'B1'), ('B5', 'C6'), ('B3', 'B5'), ('B4', 'B10'), ('A2', 'B5'), ('A2', 'C6'), ('A2', 'C5'), ('B10', 'C6'), ('B7', 'B10'), ('A1', 'B11'), ('B2', 'B7'), ('B2', 'B10'), ('B3', 'C3'), ('B3', 'B8'), ('B6', 'B11'), ('B4', 'C5'), ('B5', 'B7'), ('C1'

======================================================================================================================================================================================================================================

**Aplica el algoritmo al juego de datos generado**

Respuesta

En base a los datos aleatorios generados, se realiza el llamado a la función del algoritmo definiendo adicionalmente el valor de temperatura inicial como de 10000000.

Al contrario de lo que ocurría anteriormente, en este escenario si se estan evaluando coincidencias de horarios entre los partidos con su correpondiente penalización por coincidencia.

En estas ejecuciones tampoco fue encontrada una solución mejor que la encontrada por el algoritmo voraz.

In [7]:
horarios_referencia = horario_aleatorio
coeficientes_referencia = coeficiente_aleatorio
lista_partidos = lista_partidos_aleatorio
Temperatura_ini = 10000000

#Se aplica la función de recocido_simulado con los datos aleteatorios generados anteriormente.
solucion,df  = recocido_simulado(Temperatura_ini,lista_partidos,horarios_referencia,coeficientes_referencia)
df

La solución de referencia es:  ['S12', 'S12', 'D18', 'D16', 'D12', 'D12', 'D18', 'L20', 'V20', 'S20']
La audiencia de referencia es:  4.63725
------------------------------------------
La audiencia vecina número  1 es: 4.626875
La audiencia vecina número  9 es: 4.634125
La audiencia vecina número  11 es: 4.6591249999999995
La audiencia vecina número  18 es: 4.72725
La audiencia vecina número  19 es: 4.979875
La audiencia vecina número  21 es: 5.024875
La audiencia vecina número  34 es: 5.055
La audiencia vecina número  39 es: 5.157125
La audiencia vecina número  155 es: 5.183999999999999
La audiencia vecina número  156 es: 5.184
La mejor solución encontrada es ['S12', 'S20', 'S12', 'D18', 'L20', 'D16', 'D18', 'D12', 'D12', 'V20']
con una distancia total de 5.184


Equipo1 Equipo2 Horario  Coeficientes  Audiencia Estimada  Coef_de_corr  \
0      A1      A2     S12          0.55              1.1000          0.75   
1      B2      B8     S12          0.55              0.4950          0.75   
2      A3      B1     D12          0.45              0.5850          0.75   
3      B4     B10     D12          0.45              0.4050          0.75   
4      B9     B11     V20          0.40              0.3600          1.00   
5      B6      B7     D18          0.85              0.7650          0.75   
6      C1      C5     D18          0.85              0.3995          0.75   
7      B3      B5     L20          0.40              0.3600          1.00   
8      C3      C4     S20          1.00              0.4700          1.00   
9      C2      C6     D16          0.75              0.3525          1.00   

   Correción coincidencia  
0                0.825000  
1                0.371250  
2                0.438750  
3                0.303750  
4                0.360000  
5                0.573750  
6                0.299625  
7                0.360000  
8                0.470000  
9                0.352500

======================================================================================================================================================================================================================================

**Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo**

Respuesta

Las referencias utilizadas en el trabajo han sido:


*   Material teórico en pdf brindado por la asignatura.
*   Actividades guiadas.



======================================================================================================================================================================================================================================

**Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño.**

Respuesta

Dados los criterios simplificativos aplicados durante el trabajo, es posible que las limitaciones se hayan aplicado en exceso impidiendo la búsqueda de soluciones en zonas no permitidas.

Una mejora potencial del algoritmo diseñado podría ser liberar el espacio de búsqueda. Realizar esta acción tendría un impacto en la cantidad de potenciales soluciones que quedarían habilitadas para su estudio lo que puede justificar el uso de algoritmos evolutivos.

======================================================================================================================================================================================================================================